File: backend/tests/chain/pools.test.ts
/**
 * Tests for chain/pools.ts — the curated 3-pool summary returned by
 * `GET /api/v1/pools`.
 *
 * The listPools function calls `provider.call()` with the V2 getReserves
 * ABI and the V3 slot0 + liquidity ABIs. We stub `call()` to return
 * ABI-encoded results for each contract call.
 */
import { describe, it, expect, vi } from 'vitest';
import { Interface, getAddress } from 'ethers';
import { listPools } from '../../src/chain/pools.js';
import { POOLS, TOKENS } from '../../src/chain/addresses.js';
import { UNISWAP_V2_PAIR_ABI, UNISWAP_V3_POOL_ABI } from '../../src/chain/abis.js';
import { upstreamUnreachable } from '../../src/errors.js';

const v2Iface = new Interface(UNISWAP_V2_PAIR_ABI);
const v3Iface = new Interface(UNISWAP_V3_POOL_ABI);

/**
 * Build a stub ChainProvider where `call()` returns ABI-encoded results
 * based on the contract address and function selector. The map
 * `callMap` keys are `address:lowercase-selector-4bytes` and values
 * are the encoded result hex strings.
 */
function makeProviderWithCallMap(
  callMap: Map<string, string>,
): {
  call: ReturnType<typeof vi.fn>;
  getBlockNumber: ReturnType<typeof vi.fn>;
} {
  return {
    call: vi.fn(async (tx: { to?: string; data?: string }) => {
      if (!tx.to || !tx.data) throw new Error('no tx.to/data');
      const selector = tx.data.slice(0, 10).toLowerCase();
      const key = `${tx.to.toLowerCase()}:${selector}`;
      const result = callMap.get(key);
      if (!result) {
        throw new Error(`no stub for ${key}`);
      }
      return result;
    }),
    getBlockNumber: vi.fn(async () => 0x1234),
  };
}

describe('listPools', () => {
  it('returns 3 pools: V2 ETH/USDC + V3 ETH/USDC + V3 ETH/USDT', async () => {
    const v2Result = v2Iface.encodeFunctionResult('getReserves', [
      1000000n * 10n ** 18n, // reserve0
      2000000n * 10n ** 6n, // reserve1
      1700000000n, // blockTimestampLast
    ]);
    const slot0V3Usdc = v3Iface.encodeFunctionResult('slot0', [
      1234567890123456789n, // sqrtPriceX96
      200000n, // tick
      0, // observationIndex
      1, // observationCardinality
      1, // observationCardinalityNext
      0, // feeProtocol
      true, // unlocked
    ]);
    const liquidityV3Usdc = v3Iface.encodeFunctionResult('liquidity', [5000000n]);
    const slot0V3Usdt = v3Iface.encodeFunctionResult('slot0', [
      9876543210987654321n,
      -100000n,
      0,
      1,
      1,
      0,
      true,
    ]);
    const liquidityV3Usdt = v3Iface.encodeFunctionResult('liquidity', [8000000n]);

    const callMap = new Map<string, string>([
      [`${POOLS.V2_WETH_USDC.toLowerCase()}:0x0902f1ac`, v2Result],
      [`${POOLS.V3_WETH_USDC_3000.toLowerCase()}:0x3850c7bd`, slot0V3Usdc],
      [`${POOLS.V3_WETH_USDC_3000.toLowerCase()}:0x1a686502`, liquidityV3Usdc],
      [`${POOLS.V3_WETH_USDT_3000.toLowerCase()}:0x3850c7bd`, slot0V3Usdt],
      [`${POOLS.V3_WETH_USDT_3000.toLowerCase()}:0x1a686502`, liquidityV3Usdt],
    ]);
    const provider = makeProviderWithCallMap(callMap);

    const pools = await listPools(provider as never);

    expect(pools).toHaveLength(3);
    expect(pools.map((p) => p.protocol)).toEqual([
      'uniswap_v2',
      'uniswap_v3',
      'uniswap_v3',
    ]);
    // BigInt preserved end-to-end
    expect(typeof pools[0].reserve0).toBe('bigint');
    expect(pools[0].reserve0).toBeGreaterThan(0n);
    expect(pools[0].reserve1).toBeGreaterThan(0n);
    expect(typeof pools[1].sqrtPriceX96).toBe('bigint');
    expect(pools[1].sqrtPriceX96).toBeGreaterThan(0n);
    expect(typeof pools[1].liquidity).toBe('bigint');
    expect(pools[1].liquidity).toBeGreaterThan(0n);
  });

  it('V2 pool is checksummed and has the WETH/USDC token meta', async () => {
    const v2Result = v2Iface.encodeFunctionResult('getReserves', [1n, 2n, 0n]);
    const slot0 = v3Iface.encodeFunctionResult('slot0', [1n, 0, 0, 1, 1, 0, true]);
    const liq = v3Iface.encodeFunctionResult('liquidity', [1n]);
    const callMap = new Map<string, string>([
      [`${POOLS.V2_WETH_USDC.toLowerCase()}:0x0902f1ac`, v2Result],
      [`${POOLS.V3_WETH_USDC_3000.toLowerCase()}:0x3850c7bd`, slot0],
      [`${POOLS.V3_WETH_USDC_3000.toLowerCase()}:0x1a686502`, liq],
      [`${POOLS.V3_WETH_USDT_3000.toLowerCase()}:0x3850c7bd`, slot0],
      [`${POOLS.V3_WETH_USDT_3000.toLowerCase()}:0x1a686502`, liq],
    ]);
    const provider = makeProviderWithCallMap(callMap);
    const pools = await listPools(provider as never);
    const v2 = pools.find((p) => p.protocol === 'uniswap_v2')!;
    expect(v2.id).toBe(getAddress(POOLS.V2_WETH_USDC));
    expect(v2.token0.symbol).toMatch(/WETH|USDC/);
    expect(v2.token1.symbol).toMatch(/WETH|USDC/);
    expect(v2.token0.symbol).not.toBe(v2.token1.symbol);
  });

  it('V3 pool has feeTier = 3000 for the curated 0.3% pools', async () => {
    const v2Result = v2Iface.encodeFunctionResult('getReserves', [1n, 1n, 0n]);
    const slot0 = v3Iface.encodeFunctionResult('slot0', [1n, 0, 0, 1, 1, 0, true]);
    const liq = v3Iface.encodeFunctionResult('liquidity', [1n]);
    const callMap = new Map<string, string>([
      [`${POOLS.V2_WETH_USDC.toLowerCase()}:0x0902f1ac`, v2Result],
      [`${POOLS.V3_WETH_USDC_3000.toLowerCase()}:0x3850c7bd`, slot0],
      [`${POOLS.V3_WETH_USDC_3000.toLowerCase()}:0x1a686502`, liq],
      [`${POOLS.V3_WETH_USDT_3000.toLowerCase()}:0x3850c7bd`, slot0],
      [`${POOLS.V3_WETH_USDT_3000.toLowerCase()}:0x1a686502`, liq],
    ]);
    const provider = makeProviderWithCallMap(callMap);
    const pools = await listPools(provider as never);
    const v3 = pools.filter((p) => p.protocol === 'uniswap_v3');
    expect(v3).toHaveLength(2);
    for (const p of v3) expect(p.feeTier).toBe(3000);
  });

  it('throws UpstreamUnreachableError when V3 call fails', async () => {
    // V2 succeeds; V3 throws.
    const v2Result = v2Iface.encodeFunctionResult('getReserves', [1n, 1n, 0n]);
    const provider = {
      call: vi.fn(async (tx: { to?: string; data?: string }) => {
        if (!tx.to || !tx.data) throw new Error('no tx.to/data');
        if (tx.to.toLowerCase() === POOLS.V2_WETH_USDC.toLowerCase()) return v2Result;
        throw new Error('rpc down');
      }),
      getBlockNumber: vi.fn(async () => 0),
    };
    await expect(listPools(provider as never)).rejects.toBeInstanceOf(
      upstreamUnreachable('rpc down').constructor,
    );
    // We don't assert the exact error — only that an HttpError surfaces.
    let caught: unknown;
    try {
      await listPools(provider as never);
    } catch (e) {
      caught = e;
    }
    expect(caught).toBeDefined();
    expect((caught as { statusCode: number }).statusCode).toBe(502);
  });

  it('uses the curated token metadata, not on-chain reads', async () => {
    const v2Result = v2Iface.encodeFunctionResult('getReserves', [1n, 1n, 0n]);
    const slot0 = v3Iface.encodeFunctionResult('slot0', [1n, 0, 0, 1, 1, 0, true]);
    const liq = v3Iface.encodeFunctionResult('liquidity', [1n]);
    const callMap = new Map<string, string>([
      [`${POOLS.V2_WETH_USDC.toLowerCase()}:0x0902f1ac`, v2Result],
      [`${POOLS.V3_WETH_USDC_3000.toLowerCase()}:0x3850c7bd`, slot0],
      [`${POOLS.V3_WETH_USDC_3000.toLowerCase()}:0x1a686502`, liq],
      [`${POOLS.V3_WETH_USDT_3000.toLowerCase()}:0x3850c7bd`, slot0],
      [`${POOLS.V3_WETH_USDT_3000.toLowerCase()}:0x1a686502`, liq],
    ]);
    const provider = makeProviderWithCallMap(callMap);
    const pools = await listPools(provider as never);
    const token0Symbols = new Set(pools.flatMap((p) => [p.token0.symbol, p.token1.symbol]));
    // WETH + USDC + USDT must be present somewhere in the curated list.
    expect(token0Symbols.has('WETH')).toBe(true);
    expect(token0Symbols.has('USDC')).toBe(true);
    expect(token0Symbols.has('USDT')).toBe(true);
    // Sanity: TOKENS export agrees.
    expect(TOKENS.WETH.decimals).toBe(18);
  });
});

File: backend/tests/chain/pools_test.py
"""Unit tests for `dtm_backend.chain.pools`."""
from __future__ import annotations

import pytest

from dtm_backend.chain.pools import (
    POOL_CATALOG,
    TOKEN_METADATA,
    PoolFetcher,
)
from dtm_backend.chain.types import PoolProtocol, PoolToken


def test_token_metadata_has_five_tokens() -> None:
    """The token metadata table holds the 5 ERC20s we expose."""
    assert set(TOKEN_METADATA) == {"WETH", "USDC", "USDT", "DAI", "WBTC"}
    for symbol, token in TOKEN_METADATA.items():
        assert isinstance(token, PoolToken)
        assert token.symbol == symbol
        assert token.address.startswith("0x")
        assert 0 <= token.decimals <= 36


def test_pool_catalog_has_three_entries() -> None:
    """The pool catalog is 1 V2 pair + 2 V3 pools = 3 total.

    Preserved from the previous TypeScript backend (3 pools
    total — that is what the frontend `listPools` integration
    test expects).
    """
    assert len(POOL_CATALOG) == 3
    v2_count = sum(1 for _, proto in POOL_CATALOG if proto is PoolProtocol.UNISWAP_V2)
    v3_count = sum(1 for _, proto in POOL_CATALOG if proto is PoolProtocol.UNISWAP_V3)
    assert v2_count == 1
    assert v3_count == 2


def test_pool_catalog_addresses_are_checksummed() -> None:
    """All pool addresses in the catalog are EIP-55 checksummed."""
    from dtm_backend.chain.addresses import is_checksum_address

    for addr, _ in POOL_CATALOG:
        assert is_checksum_address(addr), f"{addr} is not EIP-55"


def _v2_call(contract_address, function_name, *args):
    async def _inner():
        if function_name == "getReserves":
            return (10**6, 10**18, 1_700_000_000)
        raise AssertionError(f"unexpected call {function_name}")
    return _inner()


def _v3_call(contract_address, function_name, *args):
    async def _inner():
        if function_name == "slot0":
            return (2**96, 200_000, 0, 1, 1, 0, True)
        if function_name == "liquidity":
            return 10**15
        if function_name == "fee":
            return 500  # V3 encodes 5 bps as 500 (in hundredths of bp)
        raise AssertionError(f"unexpected call {function_name}")
    return _inner()


def _multi_call(contract_address, function_name, *args):
    async def _inner():
        if function_name == "getReserves":
            return (10**6, 10**18, 1_700_000_000)
        if function_name == "slot0":
            return (2**96, 200_000, 0, 1, 1, 0, True)
        if function_name in {"liquidity", "totalSupply"}:
            return 10**15
        if function_name == "fee":
            return 500
        raise AssertionError(f"unexpected call {function_name}")
    return _inner()


@pytest.mark.asyncio
async def test_fetcher_with_canned_v2_data_produces_pool() -> None:
    """A V2 fetcher with canned `_call` produces a valid `Pool`."""
    f = PoolFetcher(_call=_v2_call)
    pool = await f.get_v2_pool(
        address="0x" + "1" * 40,
        token0=TOKEN_METADATA["USDC"],
        token1=TOKEN_METADATA["WETH"],
    )
    assert pool.protocol is PoolProtocol.UNISWAP_V2
    assert pool.reserve0 == 10**6
    assert pool.reserve1 == 10**18
    assert pool.token0.symbol == "USDC"
    assert pool.token1.symbol == "WETH"
    assert pool.sqrt_price_x96 is None  # V2 pools don't have slot0


@pytest.mark.asyncio
async def test_fetcher_with_canned_v3_data_produces_pool() -> None:
    """A V3 fetcher with canned `_call` produces a valid `Pool`."""
    f = PoolFetcher(_call=_v3_call)
    pool = await f.get_v3_pool(
        address="0x" + "2" * 40,
        token0=TOKEN_METADATA["USDC"],
        token1=TOKEN_METADATA["WETH"],
    )
    assert pool.protocol is PoolProtocol.UNISWAP_V3
    assert pool.sqrt_price_x96 == 2**96
    assert pool.tick == 200_000
    assert pool.fee_tier == 5
    assert pool.liquidity == 10**15


@pytest.mark.asyncio
async def test_list_all_returns_three_pools() -> None:
    """`list_all` returns 3 Pool objects when given a canned fetcher."""
    f = PoolFetcher(_call=_multi_call)
    pools = await f.list_all(POOL_CATALOG, TOKEN_METADATA)
    assert len(pools) == 3
    v2 = [p for p in pools if p.protocol is PoolProtocol.UNISWAP_V2]
    v3 = [p for p in pools if p.protocol is PoolProtocol.UNISWAP_V3]
    assert len(v2) == 1
    assert len(v3) == 2
    # V2 sanity.
    assert v2[0].reserve0 > 0
    assert v2[0].reserve1 > 0
    # V3 sanity.
    for p in v3:
        assert p.sqrt_price_x96 == 2**96
        assert p.tick == 200_000
